### PR TITLE
Review of core modern C++

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR is a review with improvements for the core modern C++ chapter. It is quite involved and can be easily split up into separate PRs if needed.

I intend to include these changes for the morning session tomorrow. But there is no need to merge it until then :)

### CODE REVIEW

File: talk/morelanguage/move.tex [hunk 1]
```diff
@@ -30,56 +28,49 @@
...
   \end{exampleblock}
   \pause
-  \begin{alertblock}{What could happen during assignment in \cpp98}
+  \begin{alertblock}{What could happen on line 4 and 6}
     \begin{itemize}
-    \item 10k allocations + 10k releases
-    \item 10k copies
+    \item one unnecessary allocation and one release for 10k \mintinline{cpp}{int}s
+    \item unnecessary copy of 10k \mintinline{cpp}{int}s
+    \item The compiler may optimize the copy away, but there is no guarantee
```
Comment (sponce) [inline talk/morelanguage/move.tex:62]:
```suggestion
    \item The compiler may optimize the copy away, but there is no guarantee (before \cpp17)
```

File: talk/morelanguage/stl.tex [hunk 2]
```diff
@@ -57,16 +56,63 @@
...
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Containers: std::vector}
+  \begin{cppcode*}{}
+    #include <vector>
+    std::vector<T> v{5, 3, 4}; // 3 Ts, 5, 3, 4
+    std::vector<T> v(100);     // 100 default constr. Ts
+    std::vector<T> v(100, 42); // 100 Ts with value 42
```
Comment (sponce) [inline talk/morelanguage/stl.tex:78]: Now we would have to explain further the notion of initializer list and why we were not really complete when saying "just use curly braces for calling constructors earlier in the course. I think we should indeed do that, but maybe in the expert course. I'll open an issue on this

File: talk/morelanguage/stl.tex [hunk 3]
```diff
@@ -136,39 +182,66 @@
...
-  \frametitlecpp[98]{STL in practice using \cpp98}
-  \begin{cppcode*}{}
-    #include <vector>
-    #include <algorithm>
+  \frametitlecpp[98]{Functors / function objects}
+  \begin{block}{Example}
+    \begin{cppcode*}{}
+      struct Incrementer {
+        int inc;
+        Incrementer(int inc) : inc(inc) {}
```
Comment (sponce) [inline talk/morelanguage/stl.tex:196]:
```suggestion
        int m_inc;
        Incrementer(int inc) : m_inc(inc) {}
```

File: talk/morelanguage/stl.tex [hunk 4]
```diff
@@ -136,39 +182,66 @@
...
 
-    std::vector<int> vi, vr(3);
-    vi.push_back(5); vi.push_back(3); vi.push_back(4);
+        int operator()(int value) const {
+          return value + m_inc;
+        };
+      };
+      std::vector<int> v{1, 2, 3};
```
Comment (sponce) [inline talk/morelanguage/stl.tex:202]:
```suggestion
      std::vector<int> v;
      v.push_back(5); v.push_back(3); ...
```
Not C++98 compatible

**Review by sponce — APPROVED** [review]
Made some tiny adjustements

File: talk/morelanguage/stl.tex [hunk 5]
```diff
@@ -136,39 +182,66 @@
...
+    \begin{cppcode*}{}
+      struct Incrementer {
+        int inc;
+        Incrementer(int inc) : inc(inc) {}
 
-    std::vector<int> vi, vr(3);
-    vi.push_back(5); vi.push_back(3); vi.push_back(4);
+        int operator()(int value) const {
+          return value + m_inc;
+        };
```
Comment (hageboeck) [inline talk/morelanguage/stl.tex:200]:
```suggestion
        }
```

File: talk/morelanguage/lambda.tex [hunk 6]
```diff
@@ -51,28 +51,30 @@
...
   \end{block}
   \begin{exampleblock}{Usage example}
     \begin{cppcode*}{firstnumber=4,gobble=2}
       std::vector<int> data{1,2,3,4,5};
-      for_each(begin(data), end(data),
-               [](int i) {
-                 std::cout << "The square of " << i
-                           << " is " << i*i << std::endl;
-               });
+      for_each(begin(data), end(data), [](int i) {
```
Comment (hageboeck) [inline talk/morelanguage/lambda.tex:67]:
```suggestion
      std::for_each(begin(data), end(data), [](int i) {
```

File: talk/morelanguage/lambda.tex [hunk 7]
```diff
@@ -191,7 +193,45 @@
...
+      };
+      auto l = __lambda4{sum, off};
+      l(42);
+    \end{cppcode*}
+  \end{multicols}
+  See also result on \href{https://cppinsights.io/lnk?code=aW50IG1haW4oKSB7CiAgaW50IHN1bSA9IDAsIG9mZnNldCA9IDE7CiAgYXV0byBsID0gWyZzdW0sIG9mZnNldF0oaW50IHgpIHsKICAgIHN1bSArPSB4ICsgb2Zmc2V0OwogIH07Cn0=&insightsOptions=cpp17&std=cpp17&rev=1.0}{cppinsights.io}.
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Higher order lambdas}
```
Comment (hageboeck) [inline talk/morelanguage/lambda.tex:234]:
```suggestion
  \frametitlecpp[11]{Higher-order lambdas}
```

File: talk/morelanguage/raii.tex [hunk 8]
```diff
@@ -126,22 +126,22 @@
...
     \end{itemize}
   \end{block}
   \pause
   \begin{exampleblock}{Usage}
     \begin{cppcode*}{}
-      Foo *p = new Foo{};  // allocation
-      std::unique_ptr<Foo> uptr(p);
-      std::cout << uptr.get() << " points to "
-                << uptr->someMember << '\n';
+      std::unique_ptr<Foo> p(new Foo{}); // allocation
```
Comment (hageboeck) [inline talk/morelanguage/raii.tex:138]:
```suggestion
      std::unique_ptr<Foo> p{ new Foo{} }; // allocation
```

**Review by hageboeck — COMMENTED** [review]
Here are my N cents.

File: talk/morelanguage/stl.tex [hunk 9]
```diff
@@ -136,39 +182,66 @@
...
 \end{frame}
 
-\againframe{STLcode}
-
 \begin{frame}[fragile]
-  \frametitlecpp[98]{STL in practice using \cpp98}
-  \begin{cppcode*}{}
-    #include <vector>
-    #include <algorithm>
+  \frametitlecpp[98]{Functors / function objects}
```
Comment (hageboeck) [inline talk/morelanguage/stl.tex:191]:
```suggestion
  \frametitlecpp[11]{Functors / function objects}
```
`Incrementer{42}` needs c++11, I guess.

File: talk/morelanguage/stl.tex (hunk 2 above)
Comment (sponce) [reply talk/morelanguage/stl.tex:78]: resolved in issue #155

**Pull request merged**